nemo_rl/algorithms/async_utils.py

```diff
@@ -0,0 +1,660 @@
+            self.trajectory_versions.append(weight_version)
+            self.target_weight_versions.append(target_weight_version)
+            self.last_target_weight_already_generated = max(
+                self.last_target_weight_already_generated, target_weight_version
+            )
+            print(
+                f"ReplayBuffer state: {len(self.trajectories)} groups, versions={self.trajectory_versions}, targets={self.target_weight_versions}, last_target_weight_already_generated={self.last_target_weight_already_generated}"
+            )
+            return "success"
+
+    def get_debug_info(self) -> dict:
+        """Get debug information about buffer state."""
+        return {
+            "total_trajectories": len(self.trajectories),
+            "trajectory_versions": self.trajectory_versions,
+            "target_weight_versions": self.target_weight_versions,
+            "max_size": self.max_size,
+        }
+
+    def get_last_target_weight_already_generated(self) -> int:
+        with self._lock:
+            return self.last_target_weight_already_generated
+
+    def get_existing_target_weights(self) -> set[int]:
+        """Get set of target weight versions that already have trajectories."""
+        with self._lock:
+            return set(self.target_weight_versions)
+
+    def sample(
+        self,
+        num_prompt_groups: int,
+        current_weight_version: int,
+        max_age_steps: int,
+    ) -> Optional[dict[str, Any]]:
+        """Sample per-prompt trajectory groups intended for the current training step.
+
+        Only returns trajectories with target_weight_version == current_weight_version.
+        If insufficient trajectories are available, returns None to stall training
+        until the remaining trajectories are generated. This ensures no trajectory
+        loses its last chance to be used for its intended training step.
+
+        Returns:
+            Dictionary with 'trajectories' and 'avg_trajectory_age' keys, or None if insufficient data
+        """
+        with self._lock:
+            if not self.trajectories:
+                return None
+
+            total_trajectories = len(self.trajectories)
+            print("🔍 ReplayBuffer sampling debug:")
+            print(
+                f"   current_weight_version={current_weight_version}, max_age_steps={max_age_steps}"
+            )
+            print(f"   trajectory_versions={self.trajectory_versions}")
+
+            # For debugging: check for unexpected old trajectories
+            from collections import Counter
+
+            version_counts = Counter(self.trajectory_versions)
+            print(f"   version_counts: {version_counts}")
+
+            # Compute minimum valid version based on age window
+            # max_age_steps=1 means trajectories from the last 1 step are valid
+            min_valid_version = max(0, current_weight_version - max_age_steps)
+            print(f"   min_valid_version={min_valid_version}")
```

Comment:
```suggestion
            print(f"   {min_valid_version=}")
```
save a few characters :)